File: arrays-and-hashing/217-contains-duplicate/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        unordered_set<int> numSet;

        for (int num : nums) {
            if (numSet.find(num) != numSet.end()) {
                return true;
            }
            numSet.insert(num);
        }
        return false;
    }
};

int main() {
    Solution solution;

    // Test case 1: Contains duplicates
    vector<int> nums1 = {1, 2, 3, 1};
    cout << boolalpha << solution.containsDuplicate(nums1) << endl;  // Output: true

    // Test case 2: No duplicates
    vector<int> nums2 = {1, 2, 3, 4};
    cout << boolalpha << solution.containsDuplicate(nums2) << endl;  // Output: false

    return 0;
}

File: arrays-and-hashing/217-contains-duplicate/solution.py
class Solution(object):
    def containsDuplicate(self, nums):
        numset = set()
        for i in nums:
            if i in numset:
                return True
            else:
                numset.add(i)
        
        return False

# More efficient solution:  
# return len(set(nums)) != len(nums)


# Explanation:

# - We could brute force the solution by duplicating the list and checking
# term by term O(n^2)

# - We can also check for duplicates by using sets
# sets cant have duplicates, converting to set removes duplicates if there are any
# if duplicates were removed, set length is different than list length